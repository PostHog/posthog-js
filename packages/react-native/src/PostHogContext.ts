import React from 'react'
import { PostHog } from './posthog-rn'

export const PostHogContext = React.createContext<{ client: PostHog }>({ client: undefined as unknown as PostHog })
