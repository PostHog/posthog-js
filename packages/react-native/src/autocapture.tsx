import { PostHog } from './posthog-rn'
import { PostHogAutocaptureElement, JsonType } from 'posthog-core'
import { PostHogAutocaptureOptions } from './types'

interface Element {
  elementType?: {
    displayName?: string
    name?: string
  }
  memoizedProps?: Record<string, unknown>
  return?: Element
}

const flattenStyles = (styles: any): any => {
  const flattened: any = {}

  if (Array.isArray(styles)) {
    for (const style of styles) {
      Object.assign(flattened, flattenStyles(style))
    }
  } else {
    Object.assign(flattened, styles)
  }

  return flattened
}

const stringifyStyle = (styles: any): string => {
  const flattened = flattenStyles(styles)

  const str = Object.keys(flattened)
    .map((x) => `${x}:${flattened[x]}`)
    .join(';')

  return str
}

const sanitiseLabel = (label: string): string => {
  return label.replace(/[^a-z0-9]+/gi, '-')
}

export const defaultPostHogLabelProp = 'ph-label'

export const autocaptureFromTouchEvent = (e: any, posthog: PostHog, options: PostHogAutocaptureOptions = {}): void => {
  const {
    noCaptureProp = 'ph-no-capture',
    customLabelProp = defaultPostHogLabelProp,
    maxElementsCaptured = 20,
    ignoreLabels = [],
    propsToCapture = ['style', 'testID', 'accessibilityLabel', customLabelProp, 'children'],
  } = options

  if (!e._targetInst) {
    return
  }
  const elements: PostHogAutocaptureElement[] = []

  let currentInst: Element | undefined = e._targetInst

  while (
    currentInst &&
    // maxComponentTreeSize will always be defined as we have a defaultProps. But ts needs a check so this is here.
    elements.length < maxElementsCaptured
  ) {
    const el: PostHogAutocaptureElement = {
      tag_name: '',
    }

    const props = currentInst.memoizedProps

    if (props?.[noCaptureProp]) {
      // Immediately ignore events if a no capture is in the chain
      return
    }

    if (props) {
      // Capture only props we have said to capture. By default this is only "safe" props
      Object.keys(props).forEach((key) => {
        if (!propsToCapture.includes(key)) {
          return
        }
        const value = props[key]
        if (key === 'style') {
          el.attr__style = stringifyStyle(value)
        } else if (['string', 'number', 'boolean'].includes(typeof value)) {
          if (key === 'children') {
            el.$el_text = typeof value === 'string' ? value : JSON.stringify(value)
          } else {
            el[`attr__${key}`] = value as JsonType
          }
        }
      })
    }

    // Try and find a sensible label
    const label =
      typeof props?.[customLabelProp] !== 'undefined'
        ? `${props[customLabelProp]}`
        : currentInst.elementType?.displayName || currentInst.elementType?.name

    if (label && !ignoreLabels.includes(label)) {
      el.tag_name = sanitiseLabel(label)
      elements.push(el)
    }

    currentInst = currentInst.return
  }

  if (elements.length) {
    // The element that was tapped, may be a child (or grandchild of an element with a customLabelProp (default: ph-label))
    // In this case, the current labels applied obscure the customLabelProp (default: ph-label)
    // To correct this, loop over the elements in reverse, and promote the customLabelProp (default: ph-label)
    const elAttrLabelKey = `attr__${customLabelProp}`
    let lastLabel: string | undefined = undefined

    for (let i = elements.length - 1; i >= 0; i--) {
      const element = elements[i]
      if (element[elAttrLabelKey]) {
        // this element had a customLabelProp (default: ph-label) set, promote it to the lastLabel
        lastLabel = element[elAttrLabelKey] as string
      }

      // if lastLabel is set, update this elements tag_name
      if (lastLabel) {
        element['tag_name'] = lastLabel
      }
    }
    posthog.autocapture('touch', elements, {
      $touch_x: e.nativeEvent.pageX,
      $touch_y: e.nativeEvent.pageY,
    })
  }
}
