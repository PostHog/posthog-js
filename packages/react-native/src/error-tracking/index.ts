import type { PostHog } from '../posthog-rn'
import { Logger, PostHogEventProperties, ErrorTracking as CoreErrorTracking } from '@posthog/core'
import { trackConsole, trackUncaughtExceptions, trackUnhandledRejections } from './utils'

type LogLevel = 'debug' | 'log' | 'info' | 'warn' | 'error'

const LogLevelList: LogLevel[] = ['debug', 'log', 'info', 'warn', 'error']

// user provided configuration
interface AutocaptureOptions {
  uncaughtExceptions?: boolean
  unhandledRejections?: boolean
  console?: boolean | LogLevel[]
}

export interface ErrorTrackingOptions {
  autocapture?: AutocaptureOptions | boolean
}

// resolved configuration
interface ResolvedAutocaptureOptions {
  uncaughtExceptions: boolean
  unhandledRejections: boolean
  console: LogLevel[]
}

interface ResolvedErrorTrackingOptions {
  autocapture: ResolvedAutocaptureOptions
}

export class ErrorTracking {
  private errorPropertiesBuilder: CoreErrorTracking.ErrorPropertiesBuilder
  private logger: Logger
  private options: ResolvedErrorTrackingOptions

  constructor(
    private instance: PostHog,
    options: ErrorTrackingOptions = {},
    logger: Logger
  ) {
    this.errorPropertiesBuilder = new CoreErrorTracking.ErrorPropertiesBuilder(
      [
        new CoreErrorTracking.PromiseRejectionEventCoercer(),
        new CoreErrorTracking.ErrorCoercer(),
        new CoreErrorTracking.ErrorEventCoercer(),
        new CoreErrorTracking.ObjectCoercer(),
        new CoreErrorTracking.StringCoercer(),
        new CoreErrorTracking.PrimitiveCoercer(),
      ],
      [CoreErrorTracking.chromeStackLineParser, CoreErrorTracking.geckoStackLineParser]
    )
    this.logger = logger.createLogger('[ErrorTracking]')
    this.options = this.resolveOptions(options)
    this.autocapture(this.options.autocapture)
  }

  captureException(input: unknown, additionalProperties: PostHogEventProperties, hint: CoreErrorTracking.EventHint) {
    try {
      const properties = this.errorPropertiesBuilder.buildFromUnknown(input, hint)
      return this.instance.capture('$exception', {
        ...properties,
        ...additionalProperties,
      } as unknown as PostHogEventProperties)
    } catch (error) {
      this.logger.error('An error occurred while capturing an $exception event:', error)
    }
  }

  private resolveOptions(options: ErrorTrackingOptions): ResolvedErrorTrackingOptions {
    const autocaptureOptions = this.resolveAutocaptureOptions(options.autocapture)
    return {
      autocapture: autocaptureOptions,
    }
  }

  private resolveAutocaptureOptions(autocapture: AutocaptureOptions | boolean = false): ResolvedAutocaptureOptions {
    if (typeof autocapture === 'boolean') {
      return {
        uncaughtExceptions: autocapture,
        unhandledRejections: autocapture,
        console: [],
      }
    }
    return {
      uncaughtExceptions: !!autocapture.uncaughtExceptions,
      unhandledRejections: !!autocapture.unhandledRejections,
      console: this.resolveConsoleOptions(autocapture.console),
    }
  }

  private resolveConsoleOptions(console: boolean | LogLevel[] = false): LogLevel[] {
    if (typeof console === 'boolean') {
      return console ? ['error'] : []
    }
    return Array.isArray(console) ? console.filter((level) => LogLevelList.includes(level)) : []
  }

  private autocaptureUncaughtErrors() {
    const onUncaughtException = (error: unknown, isFatal: boolean) => {
      const hint: CoreErrorTracking.EventHint = {
        mechanism: {
          type: 'onuncaughtexception',
          handled: false,
        },
      }
      const additionalProperties: any = {}

      if (isFatal) {
        additionalProperties['$exception_level'] = 'fatal' as CoreErrorTracking.SeverityLevel
      }

      this.captureException(error, additionalProperties, hint)

      if (isFatal) {
        void this.instance.flush().catch(() => {
          this.logger.critical('Failed to flush events')
        })
      }
    }
    try {
      trackUncaughtExceptions(onUncaughtException)
    } catch (err) {
      this.logger.warn('Failed to track uncaught exceptions: ', err)
    }
  }

  private autocaptureUnhandledRejections() {
    const onUnhandledRejection = (error: unknown) => {
      const hint: CoreErrorTracking.EventHint = {
        mechanism: {
          type: 'onunhandledrejection',
          handled: false,
        },
      }
      this.captureException(error, {}, hint)
    }

    try {
      trackUnhandledRejections(onUnhandledRejection)
    } catch (err) {
      this.logger.warn('Failed to track unhandled rejections: ', err)
    }
  }

  private autocaptureConsole(levels: LogLevel[]) {
    const onConsole = (level: LogLevel) => (error: unknown, isFatal: boolean, syntheticException?: Error) => {
      const hint: CoreErrorTracking.EventHint = {
        mechanism: {
          type: 'onconsole',
          handled: true,
        },
        syntheticException,
      }
      const additionalProperties = {
        $exception_level: level as CoreErrorTracking.SeverityLevel,
      }
      this.captureException(error, additionalProperties, hint)
    }

    try {
      for (const level of levels) {
        trackConsole(level, onConsole(level))
      }
    } catch (err) {
      this.logger.warn('Failed to track console errors: ', err)
    }
  }

  private autocapture(autocaptureOptions: ResolvedAutocaptureOptions) {
    if (autocaptureOptions.uncaughtExceptions === true) {
      this.autocaptureUncaughtErrors()
    }
    if (autocaptureOptions.unhandledRejections === true) {
      this.autocaptureUnhandledRejections()
    }
    if (autocaptureOptions.console.length > 0) {
      this.autocaptureConsole(autocaptureOptions.console)
    }
  }
}
