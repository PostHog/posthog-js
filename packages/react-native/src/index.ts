import { PostHog } from './posthog-rn'

export default PostHog
export * from './posthog-rn'
export * from './hooks/useNavigationTracker'
export * from './hooks/useFeatureFlags'
export * from './hooks/useFeatureFlag'
export * from './hooks/usePostHog'
export * from './PostHogProvider'
export * from './types'
export * from './surveys'
