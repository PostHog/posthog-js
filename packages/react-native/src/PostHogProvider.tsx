import React, { useCallback, useEffect, useMemo } from 'react'
import { GestureResponderEvent, StyleProp, View, ViewStyle } from 'react-native'
import { PostHog, PostHogOptions } from './posthog-rn'
import { autocaptureFromTouchEvent } from './autocapture'
import { useNavigationTracker } from './hooks/useNavigationTracker'
import { PostHogContext } from './PostHogContext'
import { PostHogAutocaptureOptions } from './types'
import { defaultPostHogLabelProp } from './autocapture'

export interface PostHogProviderProps {
  children: React.ReactNode
  options?: PostHogOptions
  apiKey?: string
  client?: PostHog
  autocapture?: boolean | PostHogAutocaptureOptions
  debug?: boolean
  style?: StyleProp<ViewStyle>
}

function PostHogNavigationHook({
  options,
  client,
}: {
  options?: PostHogAutocaptureOptions
  client?: PostHog
}): JSX.Element | null {
  useNavigationTracker(options?.navigation, options?.navigationRef, client)
  return null
}

export const PostHogProvider = ({
  children,
  client,
  options,
  apiKey,
  autocapture,
  style,
  debug = false,
}: PostHogProviderProps): JSX.Element | null => {
  if (!client && !apiKey) {
    throw new Error(
      'Either a PostHog client or an apiKey is required. If you want to use the PostHogProvider without a client, please provide an apiKey and the options={ disabled: true }.'
    )
  }

  const captureAll = autocapture === true
  const captureNone = autocapture === false

  const posthog = useMemo(() => {
    if (client && apiKey) {
      console.warn(
        'You have provided both a client and an apiKey to PostHogProvider. The apiKey will be ignored in favour of the client.'
      )
    }

    if (client) {
      return client
    }

    const parsedOptions = {
      ...options,
      captureAppLifecycleEvents:
        options?.captureAppLifecycleEvents !== undefined
          ? options.captureAppLifecycleEvents
          : !captureNone && captureAll,
    }

    return new PostHog(apiKey ?? '', parsedOptions)
  }, [client, apiKey]) // eslint-disable-line react-hooks/exhaustive-deps

  const autocaptureOptions = useMemo(
    () => (autocapture && typeof autocapture !== 'boolean' ? autocapture : {}),
    [autocapture]
  )

  const captureTouches = !captureNone && posthog && (captureAll || autocaptureOptions?.captureTouches)
  const captureScreens = !captureNone && posthog && (captureAll || (autocaptureOptions?.captureScreens ?? true)) // Default to true if not set
  const phLabelProp = autocaptureOptions?.customLabelProp || defaultPostHogLabelProp

  useEffect(() => {
    posthog.debug(debug)
  }, [debug, posthog])

  const onTouch = useCallback(
    (type: 'start' | 'move' | 'end', e: GestureResponderEvent) => {
      // TODO: Improve this to ensure we only capture presses and not just ends of a drag for example
      if (!captureTouches) {
        return
      }

      if (type === 'end') {
        autocaptureFromTouchEvent(e, posthog, autocaptureOptions)
      }
    },
    [captureTouches, posthog, autocaptureOptions]
  )

  return (
    <View
      {...{ [phLabelProp]: 'PostHogProvider' }} // Dynamically setting customLabelProp (default: ph-label)
      style={style || { flex: 1 }}
      onTouchEndCapture={captureTouches ? (e) => onTouch('end', e) : undefined}
    >
      <PostHogContext.Provider value={{ client: posthog }}>
        {captureScreens && <PostHogNavigationHook options={autocaptureOptions} client={posthog} />}
        {children}
      </PostHogContext.Provider>
    </View>
  )
}
