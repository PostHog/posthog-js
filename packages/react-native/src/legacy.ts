import { Platform } from 'react-native'
import { OptionalExpoApplication } from './optional/OptionalExpoApplication'
import { OptionalExpoFileSystem } from './optional/OptionalExpoFileSystem'

export const getLegacyValues = async (): Promise<{ distinctId?: string; anonymousId?: string } | undefined> => {
  // NOTE: The old react-native lib stored data in files on the filesystem.
  // This function takes care of pulling the legacy IDs to ensure we are using them if already present

  if (!OptionalExpoFileSystem || !OptionalExpoApplication) {
    return
  }

  // legacy didn't support macos, no need to check it
  if (Platform.OS === 'ios') {
    const posthogFileDirectory = `${OptionalExpoFileSystem.documentDirectory}../Library/Application%20Support/${OptionalExpoApplication.applicationId}/`
    const posthogDistinctIdFile = posthogFileDirectory + 'posthog.distinctId'
    const posthogAnonymousIdFile = posthogFileDirectory + 'posthog.anonymousId'

    const res = {
      distinctId: undefined,
      anonymousId: undefined,
    }

    try {
      res.distinctId = JSON.parse(await OptionalExpoFileSystem.readAsStringAsync(posthogDistinctIdFile))[
        'posthog.distinctId'
      ]
    } catch (e) {}

    try {
      res.anonymousId = JSON.parse(await OptionalExpoFileSystem.readAsStringAsync(posthogAnonymousIdFile))[
        'posthog.anonymousId'
      ]
    } catch (e) {}

    return res
  } else {
    // NOTE: Android is not supported here as the old SDK used a very Expo-unfriendly way of storing data
  }
}
