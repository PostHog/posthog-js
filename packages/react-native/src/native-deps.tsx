import { Platform } from 'react-native'
import { OptionalAsyncStorage } from './optional/OptionalAsyncStorage'
import { OptionalExpoApplication } from './optional/OptionalExpoApplication'
import { OptionalExpoDevice } from './optional/OptionalExpoDevice'
import { OptionalExpoFileSystem } from './optional/OptionalExpoFileSystem'
import { OptionalExpoLocalization } from './optional/OptionalExpoLocalization'
import { OptionalReactNativeDeviceInfo } from './optional/OptionalReactNativeDeviceInfo'
import { PostHogCustomAppProperties, PostHogCustomStorage } from './types'
import { OptionalReactNativeLocalize } from './optional/OptionalReactNativeLocalize'

const getDeviceType = (): string => {
  let deviceType = 'Mobile'

  if (Platform.OS === 'macos' || Platform.OS === 'windows') {
    deviceType = 'Desktop'
  } else if (Platform.OS === 'web') {
    deviceType = 'Web'
  }
  return deviceType
}

export const currentDeviceType = getDeviceType()

export const getAppProperties = (): PostHogCustomAppProperties => {
  const properties: PostHogCustomAppProperties = {
    $device_type: currentDeviceType,
  }

  if (OptionalExpoApplication) {
    properties.$app_build = OptionalExpoApplication.nativeBuildVersion
    properties.$app_name = OptionalExpoApplication.applicationName
    properties.$app_namespace = OptionalExpoApplication.applicationId
    properties.$app_version = OptionalExpoApplication.nativeApplicationVersion
  } else if (OptionalReactNativeDeviceInfo) {
    properties.$app_build = returnPropertyIfNotUnknown(OptionalReactNativeDeviceInfo.getBuildNumber())
    properties.$app_name = returnPropertyIfNotUnknown(OptionalReactNativeDeviceInfo.getApplicationName())
    properties.$app_namespace = returnPropertyIfNotUnknown(OptionalReactNativeDeviceInfo.getBundleId())
    properties.$app_version = returnPropertyIfNotUnknown(OptionalReactNativeDeviceInfo.getVersion())
  }

  if (OptionalExpoDevice) {
    properties.$device_manufacturer = OptionalExpoDevice.manufacturer
    // expo-device already maps the device model identifier to a human readable name
    properties.$device_name = OptionalExpoDevice.modelName

    // https://github.com/expo/expo/issues/6990
    // some devices return a value similar to:
    // HUAWEI/SNE-LX1/HWSNE:8.1.0/HUAWEISNE-LX1/131(C432):user/release-keys
    if (Platform.OS === 'android') {
      properties.$os_name = 'Android'
    } else {
      properties.$os_name = OptionalExpoDevice.osName
    }

    properties.$os_version = OptionalExpoDevice.osVersion
  } else if (OptionalReactNativeDeviceInfo) {
    properties.$device_manufacturer = returnPropertyIfNotUnknown(OptionalReactNativeDeviceInfo.getManufacturerSync())
    // react-native-device-info already maps the device model identifier to a human readable name
    properties.$device_name = returnPropertyIfNotUnknown(OptionalReactNativeDeviceInfo.getModel())
    properties.$os_name = returnPropertyIfNotUnknown(OptionalReactNativeDeviceInfo.getSystemName())
    properties.$os_version = returnPropertyIfNotUnknown(OptionalReactNativeDeviceInfo.getSystemVersion())
  }

  if (OptionalExpoLocalization) {
    properties.$locale = OptionalExpoLocalization.locale
    properties.$timezone = OptionalExpoLocalization.timezone
  } else if (OptionalReactNativeLocalize) {
    const localesFn = OptionalReactNativeLocalize.getLocales
    if (localesFn) {
      const locales = localesFn()

      if (locales && locales.length > 0) {
        const languageTag = locales[0].languageTag
        if (languageTag) {
          properties.$locale = languageTag
        }
      }
    }

    const timezoneFn = OptionalReactNativeLocalize.getTimeZone
    if (timezoneFn) {
      const timezone = timezoneFn()

      if (timezone) {
        properties.$timezone = timezone
      }
    }
  }

  return properties
}

// react-native-device-info returns 'unknown' if the property is not available (Web target)
const returnPropertyIfNotUnknown = (value: string | null): string | null => {
  if (value !== 'unknown') {
    return value
  }
  return null
}

export const buildOptimisiticAsyncStorage = (): PostHogCustomStorage => {
  if (OptionalExpoFileSystem) {
    const filesystem = OptionalExpoFileSystem
    return {
      async getItem(key: string) {
        const uri = (filesystem.documentDirectory || '') + key
        try {
          const stringContent = await filesystem.readAsStringAsync(uri)
          return stringContent
        } catch (e) {
          return null
        }
      },

      async setItem(key: string, value: string) {
        const uri = (filesystem.documentDirectory || '') + key
        await filesystem.writeAsStringAsync(uri, value)
      },
    }
  }

  if (OptionalAsyncStorage) {
    return OptionalAsyncStorage
  }

  throw new Error(
    'PostHog: No storage available. Please install expo-filesystem or react-native-async-storage OR implement a custom storage provider.'
  )
}
