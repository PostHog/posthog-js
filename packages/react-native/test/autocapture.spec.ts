import { autocaptureFromTouchEvent } from '../src/autocapture'

import goodEvent from './data/autocapture-event.json'
import ignoreEvent from './data/autocapture-event-no-capture.json'

describe('PostHog React Native', () => {
  jest.useRealTimers()
  describe('autocapture', () => {
    const nativeEvent = { pageX: 1, pageY: 2 }
    it('should capture a valid event', () => {
      const mockPostHog = { autocapture: jest.fn() } as any
      autocaptureFromTouchEvent({ _targetInst: goodEvent, nativeEvent }, mockPostHog)
      expect(mockPostHog.autocapture).toHaveBeenCalledTimes(1)
      expect(mockPostHog.autocapture.mock.calls[0]).toMatchSnapshot()
    })

    it('should ignore an invalid event', () => {
      const mockPostHog = { autocapture: jest.fn() } as any
      autocaptureFromTouchEvent({ _targetInst: ignoreEvent, nativeEvent }, mockPostHog)
      expect(mockPostHog.autocapture).toHaveBeenCalledTimes(0)
    })
  })
})
