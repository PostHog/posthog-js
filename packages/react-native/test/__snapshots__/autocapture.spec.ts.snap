// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostHog React Native autocapture should capture a valid event 1`] = `
[
  "touch",
  [
    {
      "$el_text": "I have the property \"ph-label\" which means touching me will be autocaptured with a specific label",
      "attr__ph-label": "special-text",
      "attr__testID": "example-ph-label",
      "tag_name": "special-text",
    },
    {
      "$el_text": "I have the property \"ph-label\" which means touching me will be autocaptured with a specific label",
      "attr__testID": "example-ph-label",
      "tag_name": "higher-level-element",
    },
    {
      "attr__style": "backgroundColor:rgba(0,0,0,.1);padding:15;borderRadius:10;marginBottom:10",
      "tag_name": "higher-level-element",
    },
    {
      "attr__ph-label": "higher-level-element",
      "tag_name": "higher-level-element",
    },
    {
      "attr__style": "flex:1",
      "tag_name": "ScrollView",
    },
    {
      "attr__style": "flex:1",
      "tag_name": "ScrollView",
    },
    {
      "tag_name": "PostHogDemoScreen",
    },
    {
      "tag_name": "StaticContainer",
    },
    {
      "tag_name": "EnsureSingleNavigator",
    },
    {
      "tag_name": "SceneView",
    },
    {
      "attr__style": "flex:1",
      "tag_name": "View",
    },
    {
      "attr__style": "flex:1;backgroundColor:rgb(242, 242, 242);flexDirection:column-reverse",
      "tag_name": "View",
    },
    {
      "attr__style": "flex:1;flexDirection:column-reverse",
      "tag_name": "Background",
    },
    {
      "tag_name": "Screen",
    },
    {
      "attr__style": "position:absolute;left:0;right:0;top:0;bottom:0;zIndex:0",
      "tag_name": "AnimatedComponent",
    },
    {
      "tag_name": "MaybeFreeze",
    },
    {
      "attr__style": "position:absolute;left:0;right:0;top:0;bottom:0;zIndex:0",
      "tag_name": "Screen",
    },
    {
      "attr__style": "position:absolute;left:0;right:0;top:0;bottom:0;zIndex:0",
      "tag_name": "MaybeScreen",
    },
    {
      "attr__style": "flex:1;overflow:hidden",
      "tag_name": "ScreenContainer",
    },
    {
      "attr__style": "flex:1;overflow:hidden",
      "tag_name": "MaybeScreenContainer",
    },
  ],
  {
    "$touch_x": 1,
    "$touch_y": 2,
  },
]
`;
